ompi/mca/osc/ucx/osc_ucx_comm.c

```diff
@@ -582,13 +818,29 @@ int ompi_osc_ucx_compare_and_swap(const void *origin_addr, const void *compare_a
     }
 
     ompi_datatype_type_size(dt, &dt_bytes);
-    ret = opal_common_ucx_wpmem_cmpswp(module->mem,*(uint64_t *)compare_addr,
-                                     *(uint64_t *)origin_addr, target,
-                                     result_addr, dt_bytes, remote_addr);
-    if (ret != OMPI_SUCCESS) {
+    if (sizeof(uint64_t) < dt_bytes) {
+        return OMPI_ERR_NOT_SUPPORTED;
+    }
+
+    uint64_t compare_val;
+    memcpy(&compare_val, compare_addr, dt_bytes);
```

Comment:
Is this to ensure 64bit alignment?

Reply:
Yeah, I've seen 64bit accesses of potentially arbitrarily aligned pointers provided by the user in several places (here and elsewhere in Open MPI, iirc). I'm not sure whether that is safe on all architectures (it is on x86 and seems to be on Arm >= v7). The C language defines it as UB if a pointer references a memory location that is not suitably aligned for the referenced type (www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf, page 74)

Reply:
There is also this nugget hidden in `opal_common_ucx_atomic_cswap`:

```C
        if (op_size == sizeof(uint64_t)) {
            *(uint64_t*)result = tmp;
        } else {
            assert(op_size == sizeof(uint32_t));
            *(uint32_t*)result = tmp;
        }
```
https://github.com/open-mpi/ompi/blob/master/opal/mca/common/ucx/common_ucx.h#L217

I can only guess that the intention there was to avoid unaligned accesses? I don't see language in the standard that prohibits `MPI_Compare_and_swap` to be called on `MPI_SHORT`.

Since we have only four relevant data type sizes to cover (1, 2, 4, 8) we could have an inlined custom `memcpy` that issues loads on appropriately typed pointers, if there is a concern about using `memcpy`.